### ISSUE TITLE
Script is resumed as soon as node-inspector is loaded

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [node-inspector/node-inspector#534]
I'm launching my program with `--debug-brk`. Execution is immediately paused. However, as soon as I load up `node-inspector`'s page in Chrome, the script resumes, and immediately crashes (with the error I'm trying to debug). This gives me no time to set, say, a breakpoint. (Incidentally, `debugger;` statements don't help.)

---

**Affects node v0.11.14 and higher. Use an earlier version of node to unblock yourself**

 - v0.11.13 (Last compatible Unstable) - http://nodejs.org/dist/v0.11.13/
 - v0.10.38 (Last compatible Stable) - http://nodejs.org/dist/v0.10.38/

Node v0.10.38 will give you the profiler 'partial support' warning which you can ignore if you do not use the profiler. Use v0.11.13 as the warning suggests if you need the profiler.

---

I'm not sure what other information I can share, here. My own code is fairly complex; if you'd like to try and reproduce this on your end, the following should approximately duplicate the environment where I'm seeing this:

    brew install nvm
    nvm install 0.11
    nvm use 0.11
    npm install -g node-inspector
    node-inspector &
    git clone git@github.com:Paws/Rulebook.git --branch letter-01
    git clone git@github.com:ELLIOTTCABLE/Paws.js.git --branch support-rulebook
    cd Paws.js
    npm install
    npm install # yes, run it twice. bugs in my codebase. unrelated.
    node --debug-brk ./Executables/paws.js check "../Rulebook/The Letters/01. On rules, meaning, and consistency.rules.yaml"
    # At this point, it should be paused ...
    # ... but as soon as you load up http://127.0.0.1:8080/debug?port=5858, for me, it resumes and crashes.

**Comment by 3y3** [comment]
Please share also error message.

**Comment by 3y3** [comment]
Can you reproduce this behavior on some siple app? (for example [node-debug-samples](https://github.com/node-inspector/node-debug-samples))

**Comment by ELLIOTTCABLE** [comment]
@3y3 from what? There's no error, it just opens, flashes the inspector for a moment, and then my code resumes-fails-exits, returning the inspector to the “Hit Cmd+O to open a file” view.

**Comment by ELLIOTTCABLE** [comment]
re: simpler codebase, I'll try that right now. (Wasn't expecting this quick of a reply at 4AM Chicago time, heh; about to crash soon. FYI.)

**Comment by ELLIOTTCABLE** [comment]
Works fine on `node-debug-samples/sourcemap`. I don't think it's a general problem with NI, I suspect it's the peculiarities of my codebase surfacing some weird edge-case. Problem is, I haven't the *slightest* idea where to start looking, because I don't … actually know how node-inspector works. >,>

**Comment by 3y3** [comment]
> Wasn't expecting this quick of a reply at 4AM Chicago time

I work by Russia's Moscow time =)

> Works fine on node-debug-samples/sourcemap

That's nice! Now we need to find that is special in your app. Some cases for test:
1. Run node-inspector without `&` this way we can confirm, that ni don't throw errors by himself (looking at his console log)
2. Run node-inspector with `--no-inject` flag, this prevents ni from modify your app code (now we wrap all console functions to trace them to frontend)
3. Try to reproduce this behavior on v0.7.4

**Comment by oliverzy** [comment]
I also have this problem but only with Node.js 0.11.15.
0.11.14 doesn't work, 0.11.13 is good

**Comment by iliakan** [comment]
I confirm the problem with 0.11.15.

**Comment by 3y3** [comment]
Thanks for targetting node version. I can confirm the probem with 0.11.15.

To @bajtos :
The current problem - now we receive `debug` event when app started with `--debug-brk` flag, previously we only check that the application is paused. But the initial problem - we can receive a `debug` event before all sources have been downloaded, in this situation [BreakEventHandler._onBreak](https://github.com/node-inspector/node-inspector/blob/master/lib/BreakEventHandler.js#L66-L68) works incorrectly.

I see two ways to fix this problem:
1) Cache all incoming events from app, before all sources will loaded in ScriptManager. We already do same thing with all incoming from frontend. In this way I prefer to emit `resource-tree` event not from `PageAgent` but from `process`.
2) Make `BreakEventHandler._onBreak` async and try to request source from app on the fly (with caching the response)

Are you see the third?

**Comment by iliakan** [comment]
Both ways look sane. Which one is easier to implement?

**Comment by bajtos** [comment]
Option 1) looks reasonable to me. It will probably fix also https://github.com/node-inspector/node-inspector/issues/314 which has been around for quite some time.

**Comment by bajtos** [comment]
> Option 1) looks reasonable to me. It will probably fix also #314 which has been around for quite some time.

On the second look, the fix of this issue will most likely not fix #314. However, we should take #314 into consideration while working on the patch, so that we don't make the situation even worse.

**Comment by 3y3** [comment]
Ok. We selected first way.

@ELLIOTTCABLE , @oliverzy , @iliakan , are you want to fix this or this is job for me?

**Comment by bajtos** [comment]
I don't think it is that easy, I'll try to get some time to work on this myself. Unfortunately that won't be sooner than in 2-3 weeks. So if there is anybody else willing to take up the challenge, then go for it!

**Comment by iliakan** [comment]
Don't treat that as a lack of support, I'm a really a user, not a developer here. Other projects eat 146% of my time.

**Comment by iliakan** [comment]
P.S. Thank you for what you're doing.

**Comment by mightyiam** [comment]
:+1:

**Comment by jonathandelgado** [comment]
+1, just updated to note 0.12.0 and this issue came up.

**Comment by jaslo** [comment]
Thanks I reported 558, and I'm glad to know someone's looking at it. At some point, I may have time to help, but not now unfortunately.

**Comment by ChrisRus** [comment]
I'm going back to 11.3 node until this sorts out. I'm dead in the water w/0.12.0 and latest node-inspector that I can't get to install due to v8-profiler issues on Debian (wheezy).

**Comment by bajtos** [comment]
@ChrisRus Try `npm install -g node-inspector@0.7`. That version has less native dependencies at the cost of missing profiler tabs.

**Comment by fiznool** [comment]
I'm in the same boat, using node 0.12 and node-inspector 0.9.1.

In case you need any further examples, I've reproduced with a very simple script. [Screencast here](http://youtu.be/qXNe3t1f5cE).

**Comment by colingoldberg** [comment]
I am having the same issue, with node 0.12.0, node-inspector 0.9.1. Typically I use node-debug for convenience, but the same thing happens running node-inspector plus the --debug-brk option.

I have been experiencing this issue with a fairly complex script. In light of the suggestions above, I have just tried it with a simple script I used to test the mathjs published module. It does not break at the first line, and runs through to completion without any ability to set a breakpoint.

ie.
node-debug test_mathjs.js

The test_mathjs.js script is below:

var math = require('/path/to/node_modules/mathjs');

var parser = math.parser();

var minutes = 1000 * 60;
var hours = minutes * 60;
var days = hours * 24;
var years = days * 365;

var str = 'now()-startTime';
var match = str.match(/[+-]/gi);
console.log(match);
if (match.length > 0) {
	var strArray = str.split(match[0]);
	console.log(strArray);
	if (strArray[0] == 'now()') {
		var expr_term1 = Date.now();
	}
	else {
		var expr_term1 = 1111111111111111;
	}
	if (strArray[1] == 'now()') {
		var expr_term2 = Date.now();
	}
	else {
		var expr_term2 = Date.parse('2015-01-28T19:18:48Z');
	}

	var scope = {
		expr_term1: expr_term1,
		expr_term2: expr_term2
	};
	var math_expr_str = 'expr_term1 ' + match[0] + ' expr_term2';
	var x = math.eval(math_expr_str, scope);
	var y = Math.round(x / days);
	//var x = parser.get('x');
	console.log(y);
}

parser.eval('x = 7 / 2');
var x = parser.get('x');
console.log(x);

var d= Date.parse('2015-01-28T19:18:48Z');

var y = Math.round(d / years);
console.log(y);

var n = Date.now();
var diff = n - d;
console.log(diff);
var y = Math.round(diff / days);
console.log(y);

>>>>>>>>>>>>>>>

Colin G

**Comment by MetaMemoryT** [comment]
So, in order to implement:

1) Cache all incoming events from app, before all sources will loaded in ScriptManager. We already do same thing with all incoming from frontend. In this way I prefer to emit resource-tree event not from PageAgent but from process.

I am trying to make these events:

```
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager addScript id: 82 localPath: C:\Users\Sean\p\websql-server\node_modules\primus\errors.js hidden? false source? true
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager _getSourceMapUrl(82) - using the suplied source
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager addScript id: 78 localPath: C:\Users\Sean\p\websql-server\bin\websql-server.js hidden? false source? true
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager _getSourceMapUrl(78) - using the suplied source
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager addScript id: 67 localPath: assert.js hidden? false source? true
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager addScript id: 77 localPath: _linklist.js hidden? false source? true
Fri, 06 Mar 2015 14:48:30 GMT node-inspector:ScriptManager addScript id: 71 localPath: _stream_writable.js hidden? false source? true
```
...

appear before which events?


I think it is these events:


```
Thu, 05 Mar 2015 23:40:30 GMT node-inspector:protocol:v8-debug request: {"seq":2,"type":"request","command":"continue","arguments":{"stepaction":"out","maxStringLength":10000}}

Thu, 05 Mar 2015 23:40:30 GMT node-inspector:protocol:v8-debug response: {"seq":3,"request_seq":2,"type":"response","command":"continue","success":true,"running":true}

```


Currently "command":"continue" is happening before addScript(s) are completed, correct?

**Comment by 3y3** [comment]
@MetaMemoryT , look at my pr https://github.com/3y3/node-inspector/commit/702542e7dbf3f05439a07760a2f37e6d59f77360

**Comment by MetaMemoryT** [comment]
Nice, I am pulling and testing your more-iojs branch.

**Comment by 3y3** [comment]
@MetaMemoryT , feedback will be useful, thanks =)

**Comment by MetaMemoryT** [comment]
so I pulled more-iojs

ran:

```
$ node bin/inspector.js

$ node --debug-brk simple-server/index.js
Debugger listening on port 5858
node.js:444
        stream = new tty.WriteStream(fd);
                 ^
TypeError: undefined is not a function
    at createWritableStdioStream (node.js:444:18)
    at process.stdout (node.js:502:16)
    at console.js:111:37
    at NativeModule.compile (node.js:805:5)
    at Function.NativeModule.require (node.js:774:18)
    at node.js:214:27
    at Server.<anonymous> (C:\Users\Sean\p\node-debug-samples\simple-server\index.js:23:3)
    at Server.g (events.js:199:16)
    at Server.emit (events.js:104:17)
    at net.js:1144:12
```

(from [node-debug-samples](https://github.com/node-inspector/node-debug-samples))

seems like another third party error?

then I ran:

```
$ node bin/inspector.js
```

```
$ node --debug-brk sourcemap/app.js
```

and still got the same problem this issue is about, and got this message in the inspector console:

TypeError: Cannot read property 'prototype' of undefined

**Comment by rjgotten** [comment]
FYI; The NodeJS tools for Visual Studio ran into this exact same issue: https://nodejstools.codeplex.com/workitem/1808

According to information from that issue, it seems that NodeJS 0.12 now signals the initial break as an actual breakpoint, and if you treat that as an unknown breakpoint and just instruct the debugger to continue, then the app immediately resumes and runs to completion.

If I'm reading this correctly, their fix is to treat unknown breakpoints differently (basically; not instruct the debugger to continue on) while the debugger hasn't finished attaching itself.

**Comment by metaford** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

z.lee@navercorp.com 님께 보내신 메일 <Re: [node-inspector] Script is resumed as soon as node-inspector is loaded (#534)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by MetaMemoryT** [comment]
Great information, @rjgotten 

Trying to find the code handling unknown breakpoints:

https://nodejstools.codeplex.com/SourceControl/latest#Nodejs/Product/Nodejs/Debugger/DebugEngine/AD7MemoryAddress.cs

https://nodejstools.codeplex.com/SourceControl/latest#Nodejs/Product/Nodejs/Debugger/DebugEngine/AD7BreakpointResolution.cs


Would this require changes to https://github.com/node-inspector/v8-debug ?

**Comment by rafis** [comment]
I have tried downgrading to node v0.11.16 and node-inspector v0.7.4 and node-inspector v0.8.3 - still can't break on first line.

**Comment by aparajita** [comment]
@rafis node-inspector 0.9.1 and node 0.11.13 is working for me.

**Comment by iliakan** [comment]
Do I get it right that this bug totally screws up debugging with `--debug-brk` for all recent iojs and node versions?

**Comment by oliverzy** [comment]
yes，0.11.13 is the last version that works

**Comment by iliakan** [comment]
Any node-inspector replacement if I still want to debug with `--debug-brk` ?

**Comment by 3y3** [comment]
@iliakan , you can try to fork from #560

**Comment by 3y3** [comment]
@iliakan , so if you'll do it, send me email, if this pr doesn't work for you correctly.

**Comment by iliakan** [comment]
@3y3 I can live without `--debug-brk` (and without `--debug` at all too! and without any dev tools! I'm an old programmer motherf*cker! :))

Just tracking the situation to jump on something when it's production ready.

**Comment by Ndrou** [comment]
it seems this bug is related to https://github.com/iojs/io.js/issues/877

**Comment by joepie91** [comment]
Experiencing the same issue. `--debug-brk` works, but as soon as I attach `node-inspector`, the code resumes running and finishes. `debugger` statements do not help, and actually segfault the process (see below). Using `node-debug` gives the same result. Figured I'd add my two cents, in case there's any useful information in it - I have a simple testcase, in any case.

I use CoffeeScript, but I am compiling with sourcemaps and then running the `.js` file instead, so the issue shouldn't be with CoffeeScript here. Exact commands run are `coffee -c -m test-leak.coffee` and `node --debug-brk test-leak.js`.

```
sven@linux-etoq:~/projects/node-bhttp> node -v
v0.12.0
sven@linux-etoq:~/projects/node-bhttp> cat /usr/lib/node_modules/node-inspector/package.json | grep version
  "version": "0.9.2",
```

CoffeeScript source:

```coffeescript
firstCode = global.code
require "./"
secondCode = global.code
console.log "Leaking global.code:", (firstCode != secondCode)
```

Compiled JS:

```javascript
// Generated by CoffeeScript 1.9.1
(function() {
  var firstCode, secondCode;

  firstCode = global.code;

  require("./");

  secondCode = global.code;

  console.log("Leaking global.code:", firstCode !== secondCode);

}).call(this);

//# sourceMappingURL=test-leak.js.map
```

---

Adding a `debugger` statement will segfault the process, upon attaching `node-inspector`.

```
sven@linux-etoq:~/projects/node-bhttp> node --debug-brk test-leak.js
Debugger listening on port 5858
Leaking global.code: true
Segmentation fault
```

CoffeeScript source:

```coffeescript
debugger;
firstCode = global.code
require "./"
secondCode = global.code
console.log "Leaking global.code:", (firstCode != secondCode)
```

Compiled JS:

```javascript
// Generated by CoffeeScript 1.9.1
(function() {
  debugger;
  var firstCode, secondCode;

  firstCode = global.code;

  require("./");

  secondCode = global.code;

  console.log("Leaking global.code:", firstCode !== secondCode);

}).call(this);

//# sourceMappingURL=test-leak.js.map
```

**Comment by zedd45** [comment]
I'm experiencing the issue (debugger / breakpoints have no effect) with Node `v0.12.0` and Node Inspector `v0.9.1`

**Comment by marchaase** [comment]
I'm also experiencing this issue, with Node v0.12.0 and Node Inspector v0.9.2

**Comment by mattkime** [comment]
i just spent a couple of hours working on this. it seems like a pretty widespread problem. something should be mentioned in the readme until its fixed.

**Comment by harindaka** [comment]
Why is all the related tickets for the same issue closed? was this fixed? I'm encountering the same issue. Please help

**Comment by mattkime** [comment]
+1 @harindaka

**Comment by bajtos** [comment]
> Why is all the related tickets for the same issue closed? 

Because there is no point in having multiple issues open for the same problem, they are just cluttering the issue list.

> was this fixed? I'm encountering the same issue. Please help

Not yet, watch this issue to track any progress.

**Comment by harindaka** [comment]
Thanks for the clarification. I confused this ticket with an already closed one. Will wait for a fix and hope for the best :D

**Comment by iliakan** [comment]
Please don't confirm the issue here. There're enough confirmations.

If another `debug-brk` related issue is closed, that's probably a dupe. 

When you write something here, all subscribed people get an email. 

The issue is important, many people are subscribed.

**Comment by dpwolfe** [comment]
Quick summary for those who land here needing a workaround (like me):

**Affects node v0.11.14 or higher.  Use an earlier version of node to unblock yourself**
- v0.11.13 (Last compatible Unstable) - http://nodejs.org/dist/v0.11.13/
- v0.10.38 (Last compatible Stable) - http://nodejs.org/dist/v0.10.38/

Node v0.10.38 will give you the profiler 'partial support' warning which you can ignore if you do not use the profiler.  Use v0.11.13 as the warning suggests if you need the profiler.

**Comment by mattkime** [comment]
@dpwolfe if your statement is accurate then it should be added to README.md

@bajtos can you weigh in?

**Comment by bajtos** [comment]
I am ok with adding the summary to README, can somebody please submit a pull request?

**Comment by williscool** [comment]
+1 :(

**Comment by 3y3** [comment]
Fixed in `0.10.0` release. Reopen this bug if the problem still persists for you.

**Comment by iliakan** [comment]
Wow! holy dude, I'll start using node-inspector once again!
It even installed on my io.js!

**Comment by mightyiam** [comment]
:joy:

**Comment by jaslo** [comment]
Thank you! Will give it a try tomorrow!

**Comment by DaHoopster** [comment]
Still experiencing the issue with node v0.12.2 and node-inspector 0.10.0. Placed breakpoint in the Express 4 app and it wasn't hit. The code just runs to completion.

**Comment by aparajita** [comment]
In my case it stopped at the first line as --debug-brk should do, but then did not stop at any breakpoints.

**Comment by williscool** [comment]
Seems to work for me now. 

I use it in cases when I'm running unit tests with `mocha --debug` and somehow it decides to skip past a breakpoint.

if I run `mocha --debug-brk` I'm now consistently getting stops using `debugger;` lines in my code.

using node v0.12.2 and node-inspector 0.10.0 with nvm 0.13.1

I will say the most consistent cases skips over debugger lines happen when running with `mocha --debug` is in a class like object instantiation with the `new` keyword.

Maybe it has something to do with jit / hotloading type of code if that helps at all.

Summary: works for me now thanks. hopefully those cases where `--debug` break for me helps you track stuff down

**Comment by 3y3** [comment]
Fixed in 0.10.1

**Comment by mightyiam** [comment]
Thank you!

**Comment by SujitKumbhar** [comment]
I am facing same problem ..i installed Ni vesion is 0.10.1
i tried with --debug-brk also

**Comment by imamtauhid** [comment]
--debug-brk not running my code, node version 0.12.2 and ni 0.11.1 here. but working with --debug

**Comment by Alex-Nabu** [comment]
windows 10 not working.

**Comment by dpwolfe** [comment]
@Alex-Nabu When you say Windows 10, I'm assuming you are using the new IE Edge browser.  Is that correct?  If so, have you tried IE (not Edge) in Windows 10 or any other browsers and had different results?

**Comment by Alex-Nabu** [comment]
@dpwolfe No I meant chome. I actually have to downgrade my nodejs version to get it working again. downgraded to 0.12.4

**Comment by sanfords** [comment]
`--debug-brk` isn't working for me at all on `Mac OS X 10.10.5` + `Node.js 0.12.7 (64-bit)` + `Chrome 45.0.2454.85 (64-bit)`.  My startup.js script is very simple:

```
debugger;

var connect = require('connect');
var http    = require('http');

var app = connect();
app.use(function(req, res) {
  console.log(req.url);
  res.end('Hello from Connect!\n');
})

http.createServer(app).listen(3000);
console.log("For status, connect your web browser to http://localhost:3000");
```

This lands me in the weeds in `module.js`.  Hitting run doesn't stop at `debugger`.

I'm running with `node --debug-brk startup.js`.  startup.js is definitely loaded in the debugger.  It doesn't stop if I put a manual breakpoint there either.

**Comment by AlexYegupov** [comment]
node --debug-brk    doesn't work (nothing is shown in node-inspector)

% node --debug-brk ./test.js                                    # 130 17:26 ~/dev/react-starter
Debugger listening on port 5858

% node-inspector --web-port=9000         # 130 17:26 ~/dev/react-starter/node_modules/bootstrap
Node Inspector v0.12.3
Visit http://127.0.0.1:9000/?ws=127.0.0.1:9000&port=5858 to start debugging.

test.js: 
--------------------------
console.log(1)
console.log(2)



node v0.12.2
Node Inspector v0.12.3

Description:    Ubuntu 14.04.3 LTS
Release:        14.04
